linera-chain/src/pending_blobs.rs

```
@@ -0,0 +1,78 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+
+use async_graphql::SimpleObject;
+use linera_base::{
+    data_types::{Blob, Round},
+    ensure,
+    identifiers::BlobId,
+};
+use linera_views::{
+    context::Context,
+    map_view::MapView,
+    register_view::RegisterView,
+    views::{ClonableView, View, ViewError},
+};
+
+use crate::ChainError;
+
+/// The pending blobs belonging to a block that can't be processed without them.
+#[derive(Debug, View, ClonableView, SimpleObject)]
+pub struct PendingBlobsView<C>
+where
+    C: Clone + Context + Send + Sync + 'static,
+{
+    /// The round in which the block is validated.
+    pub round: RegisterView<C, Round>,
+    /// Whether these blobs were already validated.
+    ///
+    /// This is only `false` for _new_ block proposals, not when re-proposing blocks from earlier
+    /// rounds or when handling validated block certificates. If it is false, the pending blobs are
+    /// only the ones published by the new block, not the ones that are only read.
+    pub validated: RegisterView<C, bool>,
+    /// The map of blobs needed to process the block.
+    pub pending_blobs: MapView<C, BlobId, Option<Blob>>,
+}
+
+impl<C> PendingBlobsView<C>
+where
+    C: Clone + Context + Send + Sync + 'static,
+{
+    pub async fn get(&self, blob_id: &BlobId) -> Result<Option<Blob>, ViewError> {
+        Ok(self.pending_blobs.get(blob_id).await?.flatten())
+    }
+
+    pub async fn maybe_insert(&mut self, blob: &Blob) -> Result<(), ViewError> {
+        let blob_id = blob.id();
+        if let Some(maybe_blob) = self.pending_blobs.get_mut(&blob_id).await? {
+            if maybe_blob.is_none() {
+                *maybe_blob = Some(blob.clone());
+            }
+        }
+        Ok(())
+    }
```

Would it be useful to signal somehow if the blob was _not_ inserted?

---

I guess so: We could then return to the client whether the blob was inserted in any of the collections. I'll do that in a follow-up PR.

---

Actually, not sure it would be useful at all:

`HandlePendingBlob` currently returns a `ChainInfo`. I'm not sure it would be more useful to replace this with a `bool`, or worth it to add another response struct that contains both a `bool` and a `ChainInfo`.

Because on the client side, there isn't necessarily an expectation that we get a `true` response: E.g. other clients might currently be sending the same validated block certificate and also upload the blobs, so we would get back `false` because now the validator already had it.

Anyway, happy to discuss this next week!

---

I was thinking more along the lines of: if we try to insert an unexpected blob, maybe we'd like to know? Even if just to log it on the validator side, because that indicates either malicious behavior or a bug. It shouldn't happen, but if it does happen, it might be better to know than just silently ignore it.